src/net/NetworkServer.cpp

```diff
@@ -0,0 +1,124 @@
+    return socket;
+}
+
+void Connection::read_loop() {
+    socket.async_read_some(boost::asio::buffer(rcvbuf),
+        boost::bind(&Connection::handle_read, shared_from_this(),
+            boost::asio::placeholders::error,
+            boost::asio::placeholders::bytes_transferred));
+}
+
+bool Connection::send(std::string message) {
+    if (!socket.is_open()) {
+        return false;
+    }
+
+    boost::system::error_code error;
+    boost::asio::write(socket, boost::asio::buffer(message), error);
+
+    // If we get these errors, it's likely a clean disconnect.
+    if ((error == boost::asio::error::eof) ||
+        (error == boost::asio::error::connection_reset) ||
+        (error == boost::asio::error::broken_pipe)) {
+        std::cerr << "[send] client disconnected.\n";
+        return false;
+    } else if (error) {
+        std::cerr << "[send] some other error: " << error << "\n";
+        return false;
+    }
+
+    return true;
+}
+
+bool Connection::read_message(std::string& message) {
+    if (message_queue.empty())
+        return false;
+    message = message_queue.pop();
+    return true;
+}
+
+void Connection::handle_read(const boost::system::error_code& error, size_t bytes_transferred) {
+    if (!error) {
+        message_queue.push(rcvbuf);
+        std::cerr << "[s] recv: " << std::string(rcvbuf);
+    } else if (error != boost::asio::error::eof) {
+        std::cerr << "FATAL handle_read error: " << error << "\n";
+        return;
+    }
+
+    // Wait for and read the next message.
+    read_loop();
+}
+
+NetworkServer::NetworkServer(boost::asio::io_service* ios, unsigned int port) :
+    io_service(ios), acceptor(*ios, tcp::endpoint(tcp::v4(), port)), next_id(0) {
+    start_accept();
+}
+
+void NetworkServer::send_to_all(std::string message) {
+    unique_lock<mutex> lock(client_list_mutex);
+
+    // No clients connected.
+    if (client_list.empty()) {
+        return;
+    }
+
+    for (auto c = client_list.begin(); c != client_list.end(); /* empty */) {
+        bool success = c->second->send(message);
+        if (!success) {
+            std::cerr << "Write failed!\n";
+            // If write failed, it's likely because of disconnect. Remove from clients.
+            c = client_list.erase(c);
+        } else {
+            c++;
+        }
+    }
+}
+
+std::map<int, std::vector<std::string>> NetworkServer::read_all_messages() {
+    std::map<int, std::vector<std::string>> messages;
+    unique_lock<mutex> lock(client_list_mutex);
+
+    // No clients connected.
+    if (client_list.empty()) {
+        return messages;
+    }
+
+    for (auto const &c : client_list) {
+        messages[c.first] = std::vector<std::string>();
+        std::string message;
+        while (c.second->read_message(message)) {
+            messages[c.first].push_back(message);
+        }
+    }
+
+    return messages;
+}
+
+void NetworkServer::start_accept() {
+    Connection::pointer new_connection =
```

Comment:
Apparently shared pointers help maintain exception safety better than normal pointers, and so lots of asio stuff uses shared pointers: http://stackoverflow.com/a/4099375/2014825